File: frontend/src/api/types/threads.ts
import type { PlayerKind } from "@/api/types/chips";

export type ThreadTopic = "question" | "hand_review" | "data_change" | "chip_request";
export type ThreadStatus = "open" | "answered" | "closed";

export interface ThreadMessage {
  id: string;
  from_manager: boolean;
  author_nickname: string | null;
  body: string | null;
  attachment_id: string | null;
  created_at: string;
}

export interface ThreadSummary {
  id: string;
  topic: ThreadTopic;
  status: ThreadStatus;
  subject: string;
  chip_request_id: string | null;
  last_message_at: string;
  last_message_preview: string | null;
  /** Игроку — непрочитанный ответ менеджера; менеджеру — непрочитанное сообщение игрока. */
  unread: boolean;
  player_id: string;
  player_nickname: string | null;
  player_kind: PlayerKind | null;
}

export interface Thread extends ThreadSummary {
  messages: ThreadMessage[];
  manager_hours: string;
}

export interface ThreadCreatePayload {
  topic: ThreadTopic;
  subject?: string;
  body: string;
  chip_request_id?: string;
}

File: backend/app/schemas/threads.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field, model_validator

from app.models.enums import PlayerKind, ThreadStatus, ThreadTopic

# История раздачи из клиента бывает длинной, но не бесконечной.
MESSAGE_MAX_LENGTH = 20_000


class ThreadMessageRead(BaseModel):
    id: UUID
    from_manager: bool
    author_nickname: str | None
    body: str | None
    attachment_id: UUID | None
    created_at: datetime


class ThreadSummary(BaseModel):
    id: UUID
    topic: ThreadTopic
    status: ThreadStatus
    subject: str
    chip_request_id: UUID | None
    last_message_at: datetime
    last_message_preview: str | None
    # Для игрока — есть ответ менеджера, который он не открывал; для менеджера — наоборот.
    unread: bool
    player_id: UUID
    player_nickname: str | None = None
    player_kind: PlayerKind | None = None


class ThreadRead(ThreadSummary):
    messages: list[ThreadMessageRead]
    # «12:00–03:00 — на связи, в другое время постараемся» (ТЗ §6.1).
    manager_hours: str


class ThreadCreate(BaseModel):
    topic: ThreadTopic = ThreadTopic.QUESTION
    subject: str | None = Field(default=None, max_length=120)
    body: str = Field(min_length=1, max_length=MESSAGE_MAX_LENGTH)
    chip_request_id: UUID | None = None

    @model_validator(mode="after")
    def chip_request_matches_topic(self) -> ThreadCreate:
        if (self.topic is ThreadTopic.CHIP_REQUEST) != (self.chip_request_id is not None):
            raise ValueError("Переписка по заявке ведётся только с указанием заявки")
        if not self.body.strip():
            raise ValueError("Сообщение пустое")
        return self


class ThreadMessageCreate(BaseModel):
    body: str = Field(min_length=1, max_length=MESSAGE_MAX_LENGTH)
